pkg/envoy/admin.go

```diff
@@ -0,0 +1,119 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package envoy
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	envoyAdmin "github.com/envoyproxy/go-control-plane/envoy/admin/v2alpha"
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gogo/protobuf/proto"
+)
+
+// Shutdown initiates a graceful shutdown of Envoy.
+func Shutdown(adminPort uint32) error {
+	_, err := doEnvoyPost("quitquitquit", "", "", adminPort)
+	return err
+}
+
+// GetServerInfo returns a structure representing a call to /server_info
+func GetServerInfo(adminPort uint32) (*envoyAdmin.ServerInfo, error) {
+	buffer, err := doEnvoyGet("server_info", adminPort)
+	if err != nil {
+		return nil, err
+	}
+
+	msg := &envoyAdmin.ServerInfo{}
+	if err := unmarshal(buffer.String(), msg); err != nil {
+		return nil, err
+	}
+
+	return msg, nil
+}
+
+// GetConfigDump polls Envoy admin port for the config dump and returns the response.
+func GetConfigDump(adminPort uint32) (*envoyAdmin.ConfigDump, error) {
```

Comment:
nit: mark as experimental. should not be used for production.

Reply:
I would think any tool that needs to perform config dump should use this (e.g. istioctl).

Reply:
Right, and these tools need to know that this method is not ready yet. Envoy will break the compatibility for this field eventually, and anything that depends on it.